python_modules/dagster/dagster_tests/asset_defs_tests/test_antlr_asset_selection.py

```diff
@@ -43,3 +46,71 @@
 def test_antlr_tree(selection_str, expected_tree_str):
     asset_selection = AntlrAssetSelection(selection_str)
     assert asset_selection.tree_str == expected_tree_str
+
+
+@pytest.mark.parametrize(
+    "selection_str",
+    ["not", "a b", "a and and", "a and", "sinks", "owner", "tag:foo=", "owner:owner@owner.com"],
+)
+def test_antlr_tree_invalid(selection_str):
+    with pytest.raises(Exception):
+        AntlrAssetSelection(selection_str)
+
+
+@pytest.mark.parametrize(
+    "selection_str, expected_assets",
+    [
+        ('"a"', AssetSelection.assets("a")),
+        ("not a", AssetSelection.all() - AssetSelection.assets("a")),
+        ("a and b", AssetSelection.assets("a") & AssetSelection.assets("b")),
+        ("a or b", AssetSelection.assets("a") | AssetSelection.assets("b")),
+        ("+a", AssetSelection.assets("a").upstream(1)),
+        ("++a", AssetSelection.assets("a").upstream(2)),
+        ("a+", AssetSelection.assets("a").downstream(1)),
+        ("a++", AssetSelection.assets("a").downstream(2)),
+        (
+            "a* and *b",
+            AssetSelection.assets("a").downstream() and AssetSelection.assets("b").upstream(),
+        ),
+        ("sinks(a)", AssetSelection.assets("a").sinks()),
+        ("roots(c)", AssetSelection.assets("c").roots()),
+        ("tag:foo", AssetSelection.tag("foo", "")),
+        ("tag:foo=bar", AssetSelection.tag("foo", "bar")),
+        ('owner:"owner@owner.com"', AssetSelection.owner("owner@owner.com")),
+        ("group:my_group", AssetSelection.groups("my_group")),
+        ("kind:my_kind", AssetSelection.tag(f"{KIND_PREFIX}my_kind", "")),
+    ],
+)
+def test_antlr_visit_basic(selection_str, expected_assets):
+    # a -> b -> c
+    @asset(tags={"foo": "bar"}, owners=["team:billing"])
+    def a(): ...
+
+    @asset(deps=[a], kinds={"python", "snowflake"})
+    def b(): ...
+
+    @asset(
+        deps=[b],
+        group_name="my_group",
+    )
+    def c(): ...
+
+    defs = [a, b, c]
+
+    assert AntlrAssetSelection(selection_str).asset_selection.resolve(
```

Comment:
Can we check for equality between the generated selection and the expected selection (rather than their resolved results)?